### ISSUE TITLE
Customer data not retrieved after executing a custom login controller

### ISSUE THREAD

**Issue by pitbulk** [magento/magento2#28428]
### Description
I customized the routes for login flow because I have a SAML SSO extension, 

I'm experiencing issues the first time I try to log the user in using the SSO flow. Customer data is not loaded.

### Preconditions
 - Magento 2.3.4 or 2.3.5
 - Custom Login controller

My login controller is similar than the [PostLogin](https://github.com/magento/magento2/blob/9544fb243d5848a497d4ea7b88e08609376ac39e/app/code/Magento/Customer/Controller/Account/LoginPost.php#L187) code 

```
$customerSession->setCustomerDataAsLoggedIn($customer);
if ($this->getCookieManager()->getCookie('mage-cache-sessid')) {
    $metadata = $this->getCookieMetadataFactory()->createCookieMetadata();
    $metadata->setPath('/');
    $this->getCookieManager()->deleteCookie('mage-cache-sessid', $metadata);
}

$resultRedirect = $this->resultRedirectFactory->create();
return $resultRedirect->setUrl($targetURL);
```

The extension also forces sections to be invalidated after my custom login, by adding a etc/frontend/sections.xml as documented on [page-caching](https://devdocs.magento.com/guides/v2.3/extension-dev-guide/cache/page-caching/private-content.html)
```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"        xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Customer:etc/sections.xsd">
    <action name="sso/saml2/acs">
        <section name="*"/>
    </action>
</config>
```

### Steps to reproduce
I implemented this simple custom login controller that will allow reproduce the issue:
https://github.com/pitbulk/magento-custom-login

If in a private browser I log in, the call to retrieve sections info does not contain the customer section.
```
http://example.com/customer/section/load/?sections=recently_viewed_product%2Crecently_compared_product
%2Cpersistent&force_new_section_timestamp=false&_=1590761846169
```

If I later log in using the normal login process and then logout and try to log again using the SSO module, the customer data is properly retrieved.
If I analyze the call done to retrieve the customer-data, it contains the customer section:
```
http://example.com/customer/section/load/?sections=recently_viewed_product%2Crecently_compared_product
%2Cpersistent%2Ccustomer%2Ccompare-products%2Clast-ordered-items
%2Ccart%2Cdirectory-data%2Ccaptcha%2Cinstant-purchase%2Creview%2Cwishlist%2Cproduct_data_storage
%2Cpaypal-billing-agreement&force_new_section_timestamp=false&_=1590761971410
```
It seems that the normal login flow set something the first time the user tries to log that is not set by my extension.

The issue I experience is similar than the one described at:
https://github.com/magento/magento2/issues/21548

The issue exists on Magento 2.3.4, 2.3.4-p2 and 2.3.5-p1

### Expected result

customer section listed on section_data_ids.
Customer data available. 

### Actual result

After enabling a private browser, login and logout using the SAML SSO flow, my sessions looks like:
```
- Cookies:
  mage-cache-sessid                                        true
  mage-cache-storage                                      %7B%7D
  mage-cache-storage-section-invalidation      %7B%7D
  private_content_version                                 b1051bef9438849ebe3f156423360282
  store                                                               default
  section_data_ids                                            %7B%22recently_viewed_product%22%3A1590766823%2C%22recently_compared_product%22%3A1590766823%2C%22persistent%22%3A1590766823%7D

- Session Storage:
 mage-cache-storage                                  {}
 mage-cache-storage-section-invalidation  {}

- Local Storage:
  mage-cache-timeout                                  "2020-05-29T16:37:08.438Z"
  mage-cache-storage-section-invalidation  {}
  mage-cache-storage                                  {"recently_viewed_product":{"count":0,"items":[],
                                                                     "data_id":1590766823},"recently_compared_product":
                                                                     {"count":0,"items":[],"data_id":1590766823},
                                                                    "persistent":{"data_id":1590766823}}
```
After enabling a private browser, login and logout using the normal flow, that allows me later to login with no issue via SAML, the sessions look like:
```
- Cookies:
  mage-cache-sessid                                        true
  mage-cache-storage                                      %7B%7D
  mage-cache-storage-section-invalidation      %7B%7D
  private_content_version                                 4366e273973e6fe629402ced87a74171
  store                                                               default
  section_data_ids                                            %7B%22customer%22%3A1590767746%2C%22compare-products%22%3A1590767746%2C%22last-ordered-items%22%3A1590767746%2C%22cart%22%3A1590767746%2C%22directory-data%22%3A1590767746%2C%22captcha%22%3A1590767746%2C%22instant-purchase%22%3A1590767746%2C%22persistent%22%3A1590767746%2C%22review%22%3A1590767746%2C%22wishlist%22%3A1590767746%2C%22recently_viewed_product%22%3A1590767746%2C%22recently_compared_product%22%3A1590767746%2C%22product_data_storage%22%3A1590767746%2C%22paypal-billing-agreement%22%3A1590767746%7D

- Session Storage:
 mage-cache-storage                                  {}
 mage-cache-storage-section-invalidation  {}

- Local Storage:
  mage-cache-timeout                                  "2020-05-29T16:54:53.287Z"
  mage-cache-storage-section-invalidation  {}
  mage-cache-storage                                  {"customer":{"data_id":1590767746},"compare-products":{"count":0,"countCaption":"0 items", ............ "persistent":{"data_id":1590767746},"review":{"nickname":"","title":"","detail":"","data_id":1590767746},"wishlist":{"counter":null,"items":[],"data_id":1590767746},"recently_viewed_product":{"count":0,"items":[],"data_id":1590767746},"recently_compared_product":{"count":0,"items":[],"data_id":1590767746},"product_data_storage":{"data_id":1590767746},"paypal-billing-agreement":{"data_id":1590767746}}
```

I removed the mage-cache-storage   and tried to login with SAML, and it worked, so it seems that
all that is required is to include the so it seems that once customer section customer on section_data_ids. I executed again the SAML SSO process and added the customer string manually inside the section_data_ids, so logged in, logged out, edited the Cookie and the process worked.

What may I do to include "customer" inside the section_data_ids?

**Comment by ksz2013** [comment]
HI, 
I am having the same issue, it is exactly as you described. I am having a customized login as well, and using this one for the 1st time the customerData is not populated at all.

**Comment by Seppo-Konttinen-Solteq** [comment]
Hi.

Exactly same problem here

**Comment by pitbulk** [comment]
@vzabaznov any help?

**Comment by Seppo-Konttinen-Solteq** [comment]
I also tried by adding

`
$customerModel = $this->customerFactory->create()->updateData($customer);
        $this->eventManager->dispatch(
            'customer_customer_authenticated',
            ['model' => $customerModel, 'password' => $password]
        );

$this->eventManager->dispatch('customer_data_object_login', ['customer' => $customer]);
`

which are in /vendor/magento/module-customer/Model/AccountManagement.php:authenticate but there was no difference.

**Comment by pitbulk** [comment]
@Seppo-Konttinen-Solteq This is a very simple module with the basics:
https://github.com/pitbulk/magento-custom-login

If we can fix it, we can later port to any sso module.

I was reviewing the login-as-customer module since it is kinda similar situation, not using the standard login. I see it already had which had a similar issue
https://github.com/magefan/module-login-as-customer/issues/13

**Comment by n1cehua1991** [comment]
@pitbulk The section may be updated only by POST or PUT form submitting or AJAX request,you action `sso/saml2/acs` is not ajax request,so section is not load.

**Comment by pitbulk** [comment]
@n1cehua1991 Thanks for replying.
The sso/saml2/acs is a POST endpoint.

The [custom login controller of the example module](https://github.com/pitbulk/magento-custom-login/blob/master/Custom/Login/Controller/Custom/Login.php#L77) is right now a GET, but will update it to be a POST.

**Comment by pitbulk** [comment]
@n1cehua1991 I [updated](https://github.com/pitbulk/magento-custom-login/commit/872dca1298220b9e40ac2daa60e48bc579dfe02c) the Custom Login controller to be processed as a POST, but the issue continues.

The [Private content doc](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/cache/page-caching/private-content.html) mentions to create a [sections.xml](https://github.com/pitbulk/magento-custom-login/blob/master/Custom/Login/etc/frontend/sections.xml) but not working.

**Comment by pitbulk** [comment]
I was debugging and I found that [customer-data.js](https://github.com/magento/magento2/blob/2.4-develop/app/code/Magento/Customer/view/frontend/web/js/customer-data.js#L95) was raising:
```
The multiplewishlist section source isnt supported.
```

If I remove that section manually in the js
```
sectionNames = sectionConfig.filterClientSideSections(sectionNames);
parameters = _.isArray(sectionNames) ? {
    sectionNames = _.without(sectionNames, "multiplewishlist");
    sections: sectionNames.join(',')
} : [];
```

The customer data is loaded using the Custom Login.

**Comment by pkarsai** [comment]
As a temporary workaround, I created a custom block that invalidates any sections that have no entry in mage-cache-storage, so they are automatically loaded:

```
<?php
 /** @var \Magento\Customer\Block\SectionConfig $block */
 ?>
 <script>
     require([
         'jquery',
         'Magento_Customer/js/section-config',
         'Magento_Customer/js/customer-data',
         'jquery/jquery-storageapi'
     ], function ($, sectionConfig, customerData) {
         var sections = <?= /* @noEscape */ $this->helper(\Magento\Framework\Json\Helper\Data::class)
                 ->jsonEncode($block->getData('sectionNamesProvider')->getSectionNames()) ?>,
             storage = $.initNamespaceStorage('mage-cache-storage').localStorage,
             invalidSections = [];

         $.each(sections, function (index, sectionName) {
             if (!storage.get(sectionName)) {
                 invalidSections.push(sectionName);
             }
         });

         if (invalidSections.length) {
             customerData.invalidate(invalidSections);
         }
     });
 </script>
```

Put this in a phtml, include it in the default.xml of the Magento_Customer module using the Magento\Customer\Block\SectionConfig as block class and you should be all set.

**Comment by complexthings** [comment]
@pitbulk had this problem too, found the issue and created a PR: https://github.com/magento/magento2/pull/29081
`*` section reloading has been broken for a while.

**Comment by sedax90** [comment]
> As a temporary workaround, I created a custom block that invalidates any sections that have no entry in mage-cache-storage, so they are automatically loaded:
> 
> ```
> <?php
>  /** @var \Magento\Customer\Block\SectionConfig $block */
>  ?>
>  <script>
>      require([
>          'jquery',
>          'Magento_Customer/js/section-config',
>          'Magento_Customer/js/customer-data',
>          'jquery/jquery-storageapi'
>      ], function ($, sectionConfig, customerData) {
>          var sections = <?= /* @noEscape */ $this->helper(\Magento\Framework\Json\Helper\Data::class)
>                  ->jsonEncode($block->getData('sectionNamesProvider')->getSectionNames()) ?>,
>              storage = $.initNamespaceStorage('mage-cache-storage').localStorage,
>              invalidSections = [];
> 
>          $.each(sections, function (index, sectionName) {
>              if (!storage.get(sectionName)) {
>                  invalidSections.push(sectionName);
>              }
>          });
> 
>          if (invalidSections.length) {
>              customerData.invalidate(invalidSections);
>          }
>      });
>  </script>
> ```
> 
> Put this in a phtml, include it in the default.xml of the Magento_Customer module using the Magento\Customer\Block\SectionConfig as block class and you should be all set.

I've tried it but it doens't works....

**Comment by pitbulk** [comment]
I confirm Magento 2.4 also experience this issue.

**Comment by gabrieldagama** [comment]
Hi @pitbulk. Thank you for your report. 
The issue has been fixed in magento/magento2#29081 by @complexthings in 2.4-develop branch
Related commit(s):
 - [4f10b5db85a2a3387af81978e616ed0d21a77c4b](https://github.com/magento/magento2/commit/4f10b5db85a2a3387af81978e616ed0d21a77c4b)
 - [1505a36a8b21c0660e1cf519b7bd703e581789c1](https://github.com/magento/magento2/commit/1505a36a8b21c0660e1cf519b7bd703e581789c1)
 - [ff99146565f414f23a4105b3f8926d31fe4a39e8](https://github.com/magento/magento2/commit/ff99146565f414f23a4105b3f8926d31fe4a39e8)
 - [8047497701d5a332ad692844f12af11f344d9854](https://github.com/magento/magento2/commit/8047497701d5a332ad692844f12af11f344d9854)
 - [dd862a21c29de46c362a050638b277c7deda34c9](https://github.com/magento/magento2/commit/dd862a21c29de46c362a050638b277c7deda34c9)
 - [cdfc6814958dc07fe47c41725b6a0715fce76309](https://github.com/magento/magento2/commit/cdfc6814958dc07fe47c41725b6a0715fce76309)
 - [c2e254c868fec04e6d356c25c6b963ea8e02518a](https://github.com/magento/magento2/commit/c2e254c868fec04e6d356c25c6b963ea8e02518a)
 - [c4900734f8608cbadd51bb1032c2a8bfad2d0870](https://github.com/magento/magento2/commit/c4900734f8608cbadd51bb1032c2a8bfad2d0870)
 - [176013fd10f724e2a33998b02c93294ec1772a40](https://github.com/magento/magento2/commit/176013fd10f724e2a33998b02c93294ec1772a40)
 - [2922c1c293e41fa1d056d868c816b6f0f2e5c160](https://github.com/magento/magento2/commit/2922c1c293e41fa1d056d868c816b6f0f2e5c160)
 - [f7da23e991e79cb4430032bdd4170ffe9440771f](https://github.com/magento/magento2/commit/f7da23e991e79cb4430032bdd4170ffe9440771f)
 - [3b397e34ae8e05fbf08b01395bd733f08cf9bb21](https://github.com/magento/magento2/commit/3b397e34ae8e05fbf08b01395bd733f08cf9bb21)
 - [6f9a2743df524fe5cd60c71eaf7db58ca828bc54](https://github.com/magento/magento2/commit/6f9a2743df524fe5cd60c71eaf7db58ca828bc54)
 - [5de71f6075d312d78dfe2509b412646665919d8f](https://github.com/magento/magento2/commit/5de71f6075d312d78dfe2509b412646665919d8f)
 - [fd4a2c7f1759d03f54faa228b66efc7af51dd5a2](https://github.com/magento/magento2/commit/fd4a2c7f1759d03f54faa228b66efc7af51dd5a2)
 - [fae1c5c48b59d2f6b46ab47deb9e1548ec248453](https://github.com/magento/magento2/commit/fae1c5c48b59d2f6b46ab47deb9e1548ec248453)

The fix will be available with the upcoming 2.4.2 release.

**Comment by pitbulk** [comment]
Glad to see it fixed. I wonder if 2.3.5.X branch could be fixed as well.
